### PR TITLE
Updated theme designer stack usage

### PR DESCRIPTION
**Pull request by micahgodbolt**
Hacking on the Theme Designer stack usage. Cleaned up a bunch of unnecessary stacks, moved sections into easier to read pieces, and updated styling to avoid fixed layouts. 

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/9347)

### CODE REVIEW

**Comment by dzearing** [comment]
Thanks for stepping up for this!

File: apps/theming-designer/src/components/Header.tsx [hunk 1]
```diff
@@ -65,64 +87,51 @@ export class Header extends React.Component<IHeaderProps, IHeaderState> {
...
+            title="Microsoft Theme Designer page"
+            aria-label="Microsoft Fabric Theme Designer page"
+            styles={pipeFabricStyles}
+          >
+            | UI Fabric Theme Designer
+          </Link>
         </Stack>
-      </Card>
+        <PrimaryButton text="Export theme" onClick={this.showPanel} />
+        <Panel
```
Comment (dzearing) [inline apps/theming-designer/src/components/Header.tsx]: While this is indeed improvement, there are a number of concerns here @aneeshack4 @jdhuntington 

Things like the Panel should be wrapped in their own pure component with minimal input. That way expanding and contracting the panel doesn't re-render the entire panel.

The Panel also has very little to do with the Header, so there is a separation of concerns here. Let the Header be the header, etc.

File: apps/theming-designer/src/components/Header.tsx [hunk 2]
```diff
@@ -65,64 +87,51 @@ export class Header extends React.Component<IHeaderProps, IHeaderState> {
...
-                  </PivotItem>
-                  <PivotItem headerText="PowerShell">
-                    <textarea className={textAreaClassName} readOnly={true} spellCheck={false} value={this.state.powershellTheme} />
-                  </PivotItem>
-                </Pivot>
-              </div>
-            </Panel>
-          </Stack>
+      <Stack horizontal verticalAlign="center" grow={0} styles={headerStackStyles}>
+        <Stack horizontal grow={1} verticalAlign="center">
```
Comment (dzearing) [inline apps/theming-designer/src/components/Header.tsx]: Do we need 2 nested stacks? There is a need to grow an area, but that feels like an abuse of use here just to get flex-grow around an area.

I built this here: https://codesandbox.io/s/w2wjrw28pw

It uses this markup:

```tsx
  <div {...p}>
    <HeaderLogoImage src="https://themingdesigner.blob.core.windows.net/$web/MicrosoftLogo.png" />
    <HeaderDivider />
    <HeaderTitle>UI Fabric Theme Designer</HeaderTitle>
    <HeaderFar>
      <DefaultButton text="Export theme" />
    </HeaderFar>
  </div>
```

File: apps/theming-designer/src/shared/Typography.tsx [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+  root: {
+    display: 'block',
+    fontWeight: 600,
+    color: theme.palette.neutralPrimary,
+    fontSize: 20,
+    marginBottom: 20
+  }
+});
+
+export const TitleText = (p: ITextProps) => <Text styles={titleTextStyles} {...p} />;
```
Comment (dzearing) [inline apps/theming-designer/src/shared/Typography.tsx]: Couple things to consider just for conversation but not a change recommendation.

You added a reference to a palette color. Was there no semantic color? There is a high risk of a contrast problem here whenever we go directly to palette.

Hardcoding font propertiesmeans they are not theme-able. For an app that's not ideal, where the design language should really be centralized in the theme. What would you have preferred here?

```
fontSize: theme.fonts.sizes.s1;
```

or a loose variable with some assumption (fontSize pulls from theme.fonts.size)

```
fontSize: 's1'
```

You used `Text` but it provided no value for you at all. If you could define the variant in the theme, would that have helped?

```
theme = {
  fonts: {
    variants: {
       title: {
         size: 20,
         weight: 600
       }
     }
  }
};

<Text variant="title" color="semanticTextColor">`
```

File: apps/theming-designer/src/components/Header.tsx [hunk 4]
```diff
@@ -25,22 +36,33 @@ const outputPanelClassName = mergeStyles({
...
 const textAreaClassName = mergeStyles({
   height: 350,
-  width: 280,
+  width: '100%',
   marginRight: 28,
   backgroundColor: 'white',
   color: '#333'
 });
 
 const microsoftLogo = mergeStyles({
```
Comment (dzearing) [inline apps/theming-designer/src/components/Header.tsx]: Observation: Direct use of mergeStyles in the code feels bad, especially when you have "styles" props which aren't.. It creates too much ambiguity around when and when not to use.

Each of these components should be married to styling with the factory function.

```
const Logo = styled('div', {
  styles: () => { ... IStyles... }
});
```

**Comment by micahgodbolt** [comment]
The inner Stack in header holds the image and link, and contains the flex grow to push the button to the far right. I could remove the stack and put grow on the link, but that would make the link the entire width of the page. 

image

File: apps/theming-designer/src/components/Header.tsx (hunk 4 above)
Comment (micahgodbolt) [reply apps/theming-designer/src/components/Header.tsx]: Yeah, i wouldn't mind cleaning all of those up eventually. Really focusing on the proper way to create a 'styledComponent' would help here.

File: apps/theming-designer/src/shared/Typography.tsx (hunk 3 above)
Comment (micahgodbolt) [reply apps/theming-designer/src/shared/Typography.tsx]: 100% on all of those. i should be better pulling from the theme values. The challenge is translating from the Figma file to the proper values in the theme. Making it easier to find the theme variable that matches the value you are trying to apply would be great. i.e. what text colors use neutralPrimary (as the hex value in the figma file was the same as neutralPrimary)

File: apps/theming-designer/src/components/Header.tsx (hunk 2 above)
Comment (micahgodbolt) [reply apps/theming-designer/src/components/Header.tsx]: I either need a stack, or a div with flex styles on it. I can't put grow on the link or buttons themselves, it has to go on a container. And the wrapper of all 3 links needs to be a stack as well. So I don't see a way past 2 stacks.

Comment (micahgodbolt) [reply apps/theming-designer/src/components/Header.tsx]: in your example, does the div have flex styles assigned to it?

**Comment by micahgodbolt** [comment]
Going to merge this in for now. Obviously way more work that we can do here. Great testbed for trying out/using our controls.

**Comment by aneeshack4** [comment]
Thank you for cleaning up the Header & making it a div & getting rid of the Card usage! And everything else - updating all the samples to use Text/etc too. Also LOVE the pivots in the slots section!!!
It looks like there are just a few things: (1) the msft logo and fabric label in the header are a little off - there shouldn't be so much space between the left side of the page - it doesn't match the main fabric site then. (2) There's a big gap between the samples card and the header so when the background's changed to a color other than white it looks a bit weird. (3) And some of the text got un-bolded. Like the fabric label next to the logo, the section headers. (4) And the pivots are not centered.
image

**Pull request merged**